crates/egui/src/text_selection/text_cursor_state.rs

```diff
@@ -300,18 +300,97 @@ pub fn find_line_start(text: &str, current_index: CCursor) -> CCursor {
     // number of multi byte chars.
     // We need to know the char index to be able to correctly set the cursor
     // later.
-    let chars_count = text.chars().count();
+    let mut char_line_indexes: Vec<usize> = create_char_line_indexes(text);
+    let line_start_index = get_line_start_index(&mut char_line_indexes, current_index.index);
 
-    let position = text
-        .chars()
-        .rev()
-        .skip(chars_count - current_index.index)
-        .position(|x| x == '\n');
+    CCursor::new(line_start_index)
+}
+
+pub fn create_char_line_indexes(text: &str) -> Vec<usize> {
+    let mut line_indexes = vec![0];
 
-    match position {
-        Some(pos) => CCursor::new(current_index.index - pos),
-        None => CCursor::new(0),
+    let mut count = 0;
+    for (_i, c) in text.chars().enumerate() {
+        count += 1;
```

Comment:
Why not just use the `i` from the enumerate? Or just remove the `enumerate()` since it is not doing anything. Or do this:

```rs
    let line_indexes: Vec<_> = text
        .lines()
        .map(|line| line.chars().count())
        .scan(0, |total, next| {
            let index = *total;
            *total += next;
            Some(index)
        })
        .collect();
```

Is the vec supposed to always contain the last char index as well? Even if it is not a new line?

Reply:
`'\r' and '\n'` should also be counted.
Saving the last index also reduces `panic or bug` occurrence when using arrays.

Reply:
Wouldn't it introduce bugs, though? You are saving the last index of the last line as if it were the first index of a new line.